manual/LowLevel/Functions.md

```diff
@@ -2,13 +2,37 @@
 
 ## Introduction
 
-TODO
+This chapter discusses how `hs-bindgen` generates bindings for C functions.
 
-## Function pointers
+## Safe vs unsafe foreign imports
 
-TODO: introduction to function pointers
+When importing a C function, GHC allows us to choose between two calling
+conventions: `safe` and `unsafe`. The distinction is important:
 
-1. For every C function, generate an additional binding for the address of that C function.
+* **Safe** foreign imports may call back into Haskell code. They are more
+  expensive.
+* **Unsafe** foreign imports may not call back into Haskell. They are faster
+  because they avoid this overhead, but using `unsafe` incorrectly can lead to
+  undefined behavior.
+
+Currently, `hs-bindgen` generates all function imports using the `safe` calling
+convention:
+
+```hs
+foreign import ccall safe "B_function_name"
+  function_name :: CInt -> IO CInt
+```
+
+This is a conservative choice that ensures correctness in all cases. While it
+means we pay the cost of the safe calling convention even for functions that
+do not need it, this cost is typically negligible unless the function is called
+in a tight loop. Users who require the performance of `unsafe` imports can
+manually wrap the generated bindings with their own `unsafe` imports after
+verifying that the C function does not call back into Haskell.
```

Comment:
This is not correct. We generate two modules: `.Safe` and `.Unsafe`; users can (must) choose whichever calling convention they want.

Reply:
Right!